### PR TITLE
[Netatmo] Adding Carbon Monoxide sensor

### PR DESCRIPTION
**Pull request by clinique**
Starting addition of the module
Resolves issue #14302 
Enhanced trace logging of requested and answers with API

### CODE REVIEW

**Comment by clinique** [comment]
NB : this PR introduces an additional scope to access Netatmo API. Existing configuration will have to regenerate their refresh-token (blaning it in UI config and redoing the auth process).

**Comment by clinique** [comment]
@lolodomo : ready for code review.

File: bundles/org.openhab.binding.netatmo/README.md [hunk 1]
```diff
@@ -642,6 +643,22 @@ All these channels are read only.
...
+
+**Supported channels for the Carbon Monoxide Detector thing:**
+
+| Channel Group | Channel Id | Item Type    | Description                                      |
+| ------------- | ---------- | ------------ | ------------------------------------------------ |
+| signal        | strength   | Number       | Signal strength (0 for no signal, 1 for weak...) |
+| signal        | value      | Number:Power | Signal strength in dBm                           |
+| timestamp     | last-seen  | DateTime     | Last time the module reported its presence       |
+| last-event    | type       | String       | Type of event                                    |
+| last-event    | time       | DateTime     | Moment of the last event for this person         |
```
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/README.md:658]: Description to be fixed.

File: bundles/org.openhab.binding.netatmo/README.md [hunk 2]
```diff
@@ -642,6 +643,22 @@ All these channels are read only.
...
+| last-event    | subtype    | String       | Sub-type of event                                |
+| last-event    | message    | String       | Last event message from this person              |
```
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/README.md:660]: Description to be fixed.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/NetatmoBindingConstants.java [hunk 3]
```diff
@@ -83,7 +83,7 @@ public class NetatmoBindingConstants {
     public static final String GROUP_DOORBELL_LAST_EVENT = GROUP_LAST_EVENT + OPTION_DOORBELL;
     public static final String GROUP_DOORBELL_SUB_EVENT = GROUP_SUB_EVENT + OPTION_DOORBELL;
     public static final String GROUP_PERSON_LAST_EVENT = GROUP_LAST_EVENT + OPTION_PERSON;
-    public static final String GROUP_SMOKE_LAST_EVENT = GROUP_LAST_EVENT + OPTION_SMOKE;
+    public static final String GROUP_ALARM_LAST_EVENT = GROUP_LAST_EVENT + OPTION_ALARM;
```
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/NetatmoBindingConstants.java:86]: Shouldn't it trigger changes in README ? New channel group last-event-alarm should appear in documentation ?
Also for smoke detector ?

File: bundles/org.openhab.binding.netatmo/README.md [hunk 4]
```diff
@@ -642,6 +643,22 @@ All these channels are read only.
...
+All these channels are read only.
+
+**Supported channels for the Carbon Monoxide Detector thing:**
+
+| Channel Group | Channel Id | Item Type    | Description                                      |
+| ------------- | ---------- | ------------ | ------------------------------------------------ |
+| signal        | strength   | Number       | Signal strength (0 for no signal, 1 for weak...) |
+| signal        | value      | Number:Power | Signal strength in dBm                           |
+| timestamp     | last-seen  | DateTime     | Last time the module reported its presence       |
+| last-event    | type       | String       | Type of event                                    |
```
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/README.md:657]: Maybe it should be last-event-alarm ?

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/data/ModuleType.java [hunk 5]
```diff
@@ -140,13 +140,16 @@ ChannelGroup.BATTERY_EXT, new ChannelGroup(Therm1ChannelHelper.class, GROUP_TYPE
             new ChannelGroup(RoomChannelHelper.class, GROUP_TYPE_ROOM_PROPERTIES, GROUP_TYPE_ROOM_TEMPERATURE),
             new ChannelGroup(SetpointChannelHelper.class, GROUP_SETPOINT)),
 
-    SMOKE_DETECTOR(FeatureArea.SECURITY, "NSD", HOME, Set.of(SmokeCapability.class, ChannelHelperCapability.class),
-            ChannelGroup.SIGNAL, ChannelGroup.TIMESTAMP,
-            new ChannelGroup(EventChannelHelper.class, GROUP_SMOKE_LAST_EVENT));
+    SMOKE_DETECTOR(FeatureArea.SECURITY, "NSD", HOME, Set.of(AlarmEventCapability.class, ChannelHelperCapability.class),
+            ChannelGroup.SIGNAL, ChannelGroup.TIMESTAMP, ChannelGroup.ALARM_LAST_EVENT),
+
+    CARBON_MONOXIDE_ALARM(FeatureArea.SECURITY, "NCO", HOME,
```
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/data/ModuleType.java:146]: Very minor: is CARBON_MONOXIDE_DETECTOR no more consistent with SMOKE_DETECTOR ?

**Comment by lolodomo** [comment]
If I am not wrong, you changed an existing channel group for the smoke detector thing.
That would be a breaking change.
Can't you use the new stuff for migrating things ? I am not sure as this binding is so particular with everything built dynamically.

File: bundles/org.openhab.binding.netatmo/README.md [hunk 6]
```diff
@@ -642,6 +643,22 @@ All these channels are read only.
 | last-event    | subtype    | String       | Sub-type of event                                |
 | last-event    | message    | String       | Last event message from this person              |
 
+### Netatmo Smart Cabon Monoxide Detector
```
[the remark this answers is not in the document]
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/README.md:646]: Corrected

File: bundles/org.openhab.binding.netatmo/README.md (hunk 1 above)
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/README.md:658]: Corrected

File: bundles/org.openhab.binding.netatmo/README.md (hunk 2 above)
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/README.md:660]: Corrected

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/NetatmoBindingConstants.java (hunk 3 above)
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/NetatmoBindingConstants.java:86]: No, this is the channel-group-type id. It is not visible in documentation nor in configuration.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/data/ModuleType.java (hunk 5 above)
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/data/ModuleType.java:146]: Ok, let's adopt CO_DETECTOR

**Comment by clinique** [comment]
> If I am not wrong, you changed an existing channel group for the smoke detector thing (even if not yet documented). That would be a breaking change. Can't you use the new stuff for migrating things ? I am not sure as this binding is so particular with everything built dynamically. You would then depend on [openhab/openhab-core#3391](https://github.com/openhab/openhab-core/pull/3391)

This PR does not introduce any breaking change.

File: bundles/org.openhab.binding.netatmo/README.md (hunk 4 above)
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/README.md:659]: This is consistent with other modules.

**Review by lolodomo — APPROVED** [review]
LGTM

**Comment by lolodomo** [comment]
> This PR does not introduce any breaking change.

If you are sure, it is safe to merge.

**Comment by lolodomo** [comment]
@clinique : please help, the binding is no more working in snapshot 3360.
```

19:35:10.937 [DEBUG] [tmo.internal.handler.ApiBridgeHandler] - Connecting to Netatmo API.
19:35:10.939 [TRACE] [tmo.internal.handler.ApiBridgeHandler] - executeUri POST  https://api.netatmo.com/oauth2/token
19:35:10.941 [TRACE] [tmo.internal.handler.ApiBridgeHandler] -  -with payload : refresh_token=xxxxx&client_secret=yyyyy&grant_type=refresh_token&client_id=zzzzz&scope=read_homecoach read_station read_thermostat write_thermostat write_camera read_presence read_carbonmonoxidedetector write_doorbell write_presence read_doorbell read_camera read_smokedetector access_camera access_presence access_doorbell
19:35:10.944 [TRACE] [tmo.internal.handler.ApiBridgeHandler] -  -with headers : Accept-Encoding: gzip, User-Agent: Jetty/9.4.50.v20221201, Content-Type: application/x-www-form-urlencoded;charset=UTF-8

19:35:11.127 [TRACE] [tmo.internal.handler.ApiBridgeHandler] -  -returned : code [400 Bad Request] body {"error":"invalid_request","error_description":"Scope is not identical as the one granted by the user"}
```
This is a problem with the new scope you added.
How can I fix that ?

**Comment by lolodomo** [comment]
> NB : this PR introduces an additional scope to access Netatmo API. Existing configuration will have to regenerate their refresh-token (blanking it in UI config and redoing the auth process).

I see that you mentioned the problem and the solution. This is certainly something to mention in the release notes.

**Comment by clinique** [comment]
Mentioned an alarm for Netatmo in openhab-distro

**Pull request merged**